src/systems/lighter_than_air_dynamics/LighterThanAirDynamics.cc

```diff
@@ -0,0 +1,464 @@
+                                  math::Matrix6d::Matrix6Corner::BOTTOM_RIGHT);
+}
+
+/////////////////////////////////////////////////
+void LighterThanAirDynamics::PreUpdate(
+      const gz::sim::UpdateInfo &_info,
+      gz::sim::EntityComponentManager &_ecm)
+{
+  if (_info.paused)
+    return;
+
+  // Get vehicle state
+  gz::sim::Link baseLink(this->dataPtr->linkEntity);
+  auto linearVelocity = _ecm.Component<components::WorldLinearVelocity>(
+                                              this->dataPtr->linkEntity);
+  auto rotationalVelocity = baseLink.WorldAngularVelocity(_ecm);
+
+  if (!linearVelocity)
+  {
+    gzerr << "no linear vel" <<"\n";
+    return;
+  }
+
+  // Transform state to local frame
+  auto pose = baseLink.WorldPose(_ecm);
+  // Since we are transforming angular and linear velocity we only care about
+  // rotation. Also this is where we apply the effects of current to the link
+  auto linear_vel = pose->Rot().Inverse() * (linearVelocity->Data());
+  auto angular_velocity = pose->Rot().Inverse() * *rotationalVelocity;
+
+  // Calculate viscous forces
+  math::Vector3d vel_eps_v = LocalVelocity(linear_vel, angular_velocity,
+                                    math::Vector3d(-this->dataPtr->epsV, 0, 0));
+
+  double q0_eps_v = DynamicPressure(vel_eps_v, this->dataPtr->airDensity);
+  double gamma_eps_v = 0.0f;
+
+  gamma_eps_v = std::atan2(std::sqrt(vel_eps_v[1]*vel_eps_v[1] +
+                                     vel_eps_v[2]*vel_eps_v[2]), vel_eps_v[0]);
+
+  double forceInviscCoeff = this->dataPtr->forceInviscCoeff;
+  double forceViscCoeff = this->dataPtr->forceViscCoeff;
+  double momentInviscCoeff = this->dataPtr->momentInviscCoeff;
+  double momentViscCoeff = this->dataPtr->momentViscCoeff;
+
+  double force_visc_mag_ = q0_eps_v*(-forceInviscCoeff*std::sin(2*gamma_eps_v)
+                + forceViscCoeff*std::sin(gamma_eps_v)*std::sin(gamma_eps_v));
+
+  double moment_visc_mag_ = q0_eps_v*(-momentInviscCoeff*std::sin(2*gamma_eps_v)
+                + momentViscCoeff*std::sin(gamma_eps_v)*std::sin(gamma_eps_v));
+
+  double visc_normal_mag_ = std::sqrt(vel_eps_v[1]*vel_eps_v[1] +
+                                      vel_eps_v[2]*vel_eps_v[2]);
+
+  double visc_normal_y_ = 0.0;
+  double visc_normal_z_ = 0.0;
+
+  if(visc_normal_mag_ > std::numeric_limits<double>::epsilon()){
+
+    visc_normal_y_ = vel_eps_v[1]/visc_normal_mag_;
+    visc_normal_z_ = vel_eps_v[2]/visc_normal_mag_;
+  }
+
+  auto force_visc = force_visc_mag_  * math::Vector3d(0,
+                                                      -visc_normal_y_,
+                                                      -visc_normal_z_);
+
+  auto moment_visc = moment_visc_mag_ * math::Vector3d(0,
+                                                      visc_normal_z_,
+                                                      -visc_normal_y_);
+
+  // Added Mass forces & Torques
+  auto force_added_mass = -1.0 * this->added_mass_force(linear_vel,
```

Comment:
IIUC, if you use the SDF `<fluid><added_mass>` tags, the added mass gets added by the physics engine. Theres no need to add the added mass terms here. I think @azeey, @mabelzhang  or @quarkytale may be able to elaborate more on that.

Reply:
Thanks for this! 

Referring to the paper, on which I have based the dynamics, the added mass effects should be incorporated in two places: (1) In the linear body acceleration and angular body acceleration, and then (2) in the coupling of the linear and angular velocities. I have added picture from the paper. 

From my understanding, the physics engine incorporates it into the former (the acceleration terms). I have just added the forces and torques which is created by the added mass due to the coupling of the linear and angular velocities. I hope this make sense. 



![image](https://github.com/gazebosim/gz-sim/assets/24815968/54007818-8e6e-446a-b1ce-34e122ce9cb4)
Equation 1
![image](https://github.com/gazebosim/gz-sim/assets/24815968/05dab283-385a-49c3-9a5d-de08f85c35cd)

Reply:
Would be awesome if someone could provide insight into the backend of the physics engine and how it incorporate the added mass effects. From my understanding, it would only compute the first term in equation (6) and not both terms.

Reply:
Yep! You are right. The physics engine will only include added mass in the acceleration term. Any interaction due to velocities would have to be done outside of the engine.